extensions/amp-fit-text/1.0/fit-text.js

```diff
@@ -0,0 +1,152 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import * as styles from './fit-text.css';
+import {px, setStyle, setStyles} from '../../../src/style';
+import {useCallback, useLayoutEffect, useRef} from '../../../src/preact';
+
+const {LINE_HEIGHT_EM_} = styles;
+
+/**
+ * @param {!JsonObject} props
+ * @return {PreactDef.Renderable}
+ */
+export function FitText(props) {
+  const {
+    'children': children,
+    'minFontSize': minFontSize = 6,
+    'maxFontSize': maxFontSize = 72,
+    'width': width = props['style']['width'] || '100%',
+    'height': height = props['style']['height'] || '100%',
+    ...rest
+  } = props;
+  const contentRef = useRef(null);
+  const measurerRef = useRef(null);
+
+  const resize = useCallback(
+    (maxHeight, maxWidth) => {
+      if (!measurerRef.current) {
+        return;
+      }
+      const fontSize = calculateFontSize(
+        measurerRef.current,
+        Number(maxHeight),
+        Number(maxWidth),
+        minFontSize,
+        maxFontSize
+      );
+      getOverflowStyle(measurerRef.current, Number(maxHeight), fontSize);
+    },
+    [maxFontSize, minFontSize]
+  );
+
+  // Here and below, useLayoutEffect is used so intermediary font sizes
+  // during resizing are resolved before the component visually updates.
+  // Font size should readjust when container resizes.
+  useLayoutEffect(() => {
+    const node = contentRef.current;
+    if (!node) {
+      return;
+    }
+    const observer = new ResizeObserver((entries) => {
+      const last = entries[entries.length - 1];
+      const {height: maxHeight, width: maxWidth} = last['contentRect'];
+      resize(maxHeight, maxWidth);
+    });
+    observer.observe(node);
+    return () => observer.disconnect();
+  }, [resize]);
+
+  // Font size should readjust when content changes.
+  useLayoutEffect(() => {
```

Comment:
This one we get for free on the first resize.

Reply:
Yes that's right, are you suggesting fine tuning more?

Reply:
I meant that we don't need `useLayoutEffect` here, do we? If we don't have it, then we also don't need to pass width/height from the outside.

Reply:
Changed this to `useEffect` and moved the `width` and `height` deps to the `useCallback` deps for the `resize` method.

Reply:
Took another look at this, and with `useEffect` instead of `useLayoutEffect` you can see some intermediary styling while the font size is being calculated. I'd prefer to keep `useLayoutEffect` since its a bit more polished.

`useEffect`: 
![image](https://user-images.githubusercontent.com/10456171/81826885-3af99900-9506-11ea-9739-b7109c70565a.png)

`useLayoutEffect`:
![image](https://user-images.githubusercontent.com/10456171/81826979-51075980-9506-11ea-802c-6abd13adb152.png)

Reply:
I agree with `useLayoutEffect` vs `useEffect`. However, here are two important nuances:

1. `children` always almost change on re-render, even if they are in fact didn't change. That's how Preact works - they don't want to do item-by-item comparisons on the `children` array. So this callback will trigger on each re-render and could be wasteful.
2. The opposite is also true: `children` do not need to change for the size to change. E.g. a parent font change could lead to children resize. We could potentially ignore this for awhile, but if we'd get it for free from resize observer, that'd be potentially a good win.

It's fine to defer this as a TODO in the tracker. But it's something that we should deal with very soon.

Reply:
We get (2) with the `resize` callback that is dependent on the min and max font size props, no?

I'll add (1) to the open tasks.

Reply:
(2): Not always. Min/max or any other props do not need to change. It's enough for some far away parent to change a font-family which may never even be propagated this far.

Reply:
That being said, I'm ok if this is an open task as well.